### ISSUE TITLE
argoCD does not deploy helm dependencies 

### ISSUE THREAD

**Issue by srehmanproov** [argoproj/argo-cd#3451]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a
question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [ ] I've searched in the docs and FAQ for my answer: http://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

Defining dependencies with values.yaml file for app, argoCD does not deploy dependency. With helm only dependency is pulled down and installed.
Charts.yaml for application is:
```
dependencies:
- name: mongodb-replicaset
    version: 3.9.5
    repository: "@stable"
    alias: mongo
    condition: mongo.enabled
    ========
    value file has:
    mongo:
  image:
    tag: "4.0"
  replicas: 3
  metrics:
    enabled: true
    prometheusServiceDiscovery: true
  extraLabels:
    openvpn: inbound
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - topologyKey: "kubernetes.io/hostname"
          labelSelector:
            matchLabels:
              app: mongo 
```

**To Reproduce**

Create a app and chart.yaml define mongoDB as dependency

**Expected behavior**

Instance of mongoDB is installed together with application

**Screenshots**

NA

**Version**

```shell
argocd: v1.4.2+48cced9
  BuildDate: 2020-01-24T01:07:43Z
  GitCommit: 48cced9d925b5bc94f6aa9fa4a8a19b2a59e128a
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v1.4.2+48cced9
  BuildDate: 2020-01-24T01:07:03Z
  GitCommit: 48cced9d925b5bc94f6aa9fa4a8a19b2a59e128a
  GitTreeState: clean
  GoVersion: go1.12.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: Version: {Version:kustomize/v3.2.1 GitCommit:d89b448c745937f0cf1936162f26a5aac688f840 BuildDate:2019-09-27T00:10:52Z GoOs:linux GoArch:amd64}
  Helm Version: v2.15.2
  Kubectl Version: v1.14.0.
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by Gallardo994** [comment]
Bumping. Unable to install bitnami/mongodb chart: https://github.com/bitnami/charts

Logs I'm getting: 
```
Unable to create application: application spec is invalid: InvalidSpecError: Unable to generate manifests in bitnami/mongodb: rpc error: code = Unknown desc = 'helm2 dependency build' failed exit status 1: Error: requirements.lock is out of sync with requirements.yaml
```

**Comment by L-U-C-K-Y** [comment]
Joining in, I'm facing issues figuring out how to add remote charts and private values.yaml files.

**Comment by Relers** [comment]
same here, need to be fixed

**Comment by dansou901** [comment]
Same here. Works only if you do 'helm dependency update' in your chart folder before uploading stuff. Is not ideal because you have to push the chart to the git repository containing your app, but works.

**Comment by KlavsKlavsen** [comment]
So totally guessing (and not an argo-cd dev :) - argo-cd doesn't run "helm dependency update" before running helm template it seems? (and hence works on outdated dependency in charts/ subfolder)

**Comment by dansou901** [comment]
@KlavsKlavsen correct, seems to be that way.That is also why uploading the charts/ subfolder along with an argocd project works. Shouldn't be necessary though, you normally don't want to pollute your git repos with tar.gz archives.

**Comment by KlavsKlavsen** [comment]
@dansou901 So someone who can replicate this problem in a development setup - could probably rather easily:
1) clone argo-cd repo, 
2) find the place this happens (git grep "helm template"..)
3) add 'helm dependency update' right before that.. 
4) build updated docker image of argo-cd
5) point development argocd to the built docker and see if "problem is resolved" 
? :)

**Comment by dansou901** [comment]
Should be like that, yes.

**Comment by KlavsKlavsen** [comment]
I just hit this problem myself.. I don't need to git add charts folder though - so I put that in .gitignore.

**Comment by aritas1** [comment]
`rpc error: code = Unknown desc = 'helm dependency build' failed exit status 1: Error: no repository definition for https://helm.releases.hashicorp.com. Please add the missing repos via 'helm repo add'` with argo 1.8.4

this is a problem for me with automated patches/updates via renovate bot. the bot will upgrade the helm files but pushing the tar.gz is not an option.

**Comment by KlavsKlavsen** [comment]
@aritas1 you don't have to push the charts/ folder (which contains the .tar.gz) - you just have to push the Charts.lock file

**Comment by aritas1** [comment]
@KlavsKlavsen not for me. i removed the charts folder added it to the gitignore and still the same problem.

did you add a chart from the "official"/defalut helm registry as a dependency?
what helm version are you targeting? helm v2 oder helm v3?

**Comment by KlavsKlavsen** [comment]
I'm using application: v2 - which is helm 3.
Following: https://www.arthurkoziel.com/setting-up-argocd-with-helm/

**Comment by aritas1** [comment]
@KlavsKlavsen thanks for sharing. I found out why it's working for your tutorial setup, see the helm repositories config here:
https://github.com/arthurk/argocd-example-install/blob/master/charts/argo-cd/values.yaml

I added the hashicorp repo to my argo cd config and it's working now.

this may be feasible for some setups. the deploying user must have access to the core config of argo cd and must be aware that he/she has to add the dependencies also at the argo cd level.

**Comment by KlavsKlavsen** [comment]
@aritas1 Normally you'd want to centrally specify WHICH repos you trust (thats why there's a central repo list) - and then those repos are the only ones your users can reference in their applications. Its definetely a good security measure in my view.

**Comment by bitdiscovery** [comment]
I'm having a similar issue but with dependencies of sub-charts. My app is structures like this:

```
umbrella chart
   +--- component chart
       +---- mongodb dependency
```

What I observe is that ArgoCD thinks all is green and up-to-date, but the mongodb manifests are simply no there. Checking in the charts directory, works around it.

**Comment by killsto** [comment]
I ran into this too. I found that ArgoCD will deploy the dependencies if the dependencies are added to requirements.yaml instead of chart.yaml. So this is still a bug but it isn't as annoying as it looks at first.

Hopefully this saves someone the time I spent looking for a more complex workaround.

**Comment by KlavsKlavsen** [comment]
@killsto I believe that depends on application version of your application? (ie. which helm version you tell it to use. v1 = helm 2, v2 = helm 3). It installs dependencies fine with them listed in Chart.yaml - as long as I have done "helm dep up" on the chart and added the Chart.lock file only (ignoring charts/ folder).

**Comment by khaled** [comment]
Also running into this with a helm v3 chart (`apiVersion: v2`).  I get the following error:
```
`helm dependency build` failed exit status 1: Error: no repository definition for https://raw.githubusercontent.com/jenkinsci/kubernetes-operator/master/chart. Please add the missing repos via 'helm repo add'
```

I tried adding the dependency's helm repo as an argocd repo, and it did not help.  I don't believe this should be required anyway, as (a) the dependency's helm repository does not require credentials and (b) charts can be installed from other helm repos (via argo Applications) just fine.

**Comment by KlavsKlavsen** [comment]
you ONLY need to add the Chart.lock file - and that protects you against supply-chain attacks (on the helm repo you're pointing to) - as the hash should be checked on every run in argocd - hence it will fail (and not run) if the same version of chart is suddenly altered (which really should ONLY happen in a supply chain attack).
The error you state just means you need to add the helm repository to your argocd list of repositories in general config.

**Comment by KlavsKlavsen** [comment]
@khaled you seem to be pointing directly at a github repo.. helm repos does not suport that AFAIK

**Comment by khaled** [comment]
@KlavsKlavsen not true - it's a valid helm repo.  `helm depedency build|update` work just fine, and the chart installs successfully directly via `helm`.

**Comment by KlavsKlavsen** [comment]
Then I have no idea whats up - it works perfectly for me with 1.8.4, doing as described. Whenever I have get an error - it also shows the helm command run - and using that exact command (which has a LOT of options) - I can normally always reproduce the issue locally. Try that?

**Comment by khaled** [comment]
The error message shared in my original comment says that `helm dependency build` failed, which works just fine for me locally.

**Comment by KlavsKlavsen** [comment]
Lets hope someone can tell you what command they run exactly, so you can see if you can reproduce.. or perhaps debug can be disabled so logging is more verbose about whats wrong here.

**Comment by KarstenSiemer** [comment]
@khaled maybe update your local helm version
I was confused in the same way like you are until I updated my local helm to 3.5.3
Then I see the same error.

Not sure if I like that behavior though.
Why shouldn't someone be able to use any openly available repo for their argocd app?
If you are afraid of someone pointing to a malicious repo aren't you trying to fix the problem at the wrong place?
There are multiple ways to deploy malicious code. Someone trying to do that shouldn't be able to deploy an app at all.
Why could they auth to argocd? How did they auth to your git instance?
Like why were they able to commit to your git repo to point somewhere in the first place?
It just feels weird to me that I have to tell my colleagues to wait for me to update the argocd-cm until they can deploy some chart.

Adding either a Chart.lock nor a requirements.lock fixed the issue for me.
Likewise argocd and helm fail to deploy using `helm dependency build`.
```
Error: no repository definition for https://charts.bitnami.com/bitnami, https://charts.bitnami.com/bitnami, https://charts.bitnami.com/bitnami. Please add the missing repos via 'helm repo add'
```

But `helm dependency up` works for helm and I'd argue that argocd should do the same.
I could just locally pull whatever chart and helm template myself and point argocd to that if I want to deploy something malicious. The aspect of security just doesn't make any sense to me in this case like is argocd also preventing using kustomize remote bases? No, it is not. Then it should not behave like that with helm either imo.

**Comment by KlavsKlavsen** [comment]
> If you are afraid of someone pointing to a malicious repo aren't you trying to fix the problem at the wrong place

Well.. IF Helm charts had GPG signatures on them, so helm client could verify the dl'ed chart matches the GPG signature (as apt/yum/dnf do with deb/rpm packages) then yes - but until thats the case - I like that I can LOCK the hash, og the chart I have verified in my git commit - when I'm doing something as insecure as pointing to a public repo - for pulling down chart everytime argocd needs it.

**Comment by KarstenSiemer** [comment]
hmmm... I see. That's true though.

**Comment by musabmasood** [comment]
I guess I'm in the same boat  😢 

https://github.com/argoproj/argo-cd/issues/5821

**Comment by KlavsKlavsen** [comment]
I just managed to reproduce this by setting limits on reposerver (following the defaults in values.yaml) - it suddenly had to re-check sync on all applications (because it does not save state and got OOM killed) - in #5903 
Now trying to fix - by killing the repoServer pods that fails on this (and get stuck).. hoping that works.

**Comment by KlavsKlavsen** [comment]
``` time="2021-03-30T08:53:23Z" level=info msg="manifest error cache hit: &ApplicationSource{RepoURL:https://github.com/examplek8s-init.git,Path:argocd-helm-charts/argo-cd,TargetRevision:HEAD,Helm:&ApplicationSourceHelm{Value │
│ Files:[values.yaml ../../argocd-clusters-managed/example/values-argo-cd.yaml],Parameters:[]HelmParameter{},ReleaseName:,Values:,FileParameters:[]HelmFileParameter{},Version:,},Kustomize:nil,Ksonnet:nil,Director │
│ y:nil,Plugin:nil,Chart:,}/fb4125753e591f3ab12ed92d6369efb958d9d24e"                                                                                                                                                                   │
│ time="2021-03-30T08:53:23Z" level=error msg="finished unary call with code Unknown" error="Manifest generation error (cached): `helm dependency build` failed signal: killed" grpc.code=Unknown grpc.method=GenerateManifest grpc.req │
│ uest.deadline="2021-03-30T08:54:23Z" grpc.service=repository.RepoServerService grpc.start_time="2021-03-30T08:53:23Z" grpc.time_ms=291.198 span.kind=server system=grpc  
```
No hint as to WHY it was killed :(

**Comment by KarstenSiemer** [comment]
I resorted to helm templating all argocd helm charts and now pick them up via kustomize. Thats way less hassle in my opinion. I used [this](https://github.com/syncier/helmt) wrapper for helm template to make updating the charts easier

**Comment by KarstenSiemer** [comment]
@KlavsKlavsen using helm in argocd uses quite a lot of memory. Maybe it was just oomed either by your resource limit or by the node it is running on?

**Comment by KlavsKlavsen** [comment]
I tried even removing memory limit and this persists.. It seems to be some "perhaps node-wide"? cache inconsistency.. argocd is totally stuck.. killing pods/changing replica count.. does not help as it has in the past :(

**Comment by KlavsKlavsen** [comment]
I tried setting repoServer replicas to 0, doing docker system prune --volumes on k8s nodes.. and then starting repoServer up with1 replica.. Didn't fix a thing - same error as above :(

**Comment by KlavsKlavsen** [comment]
> I resorted to helm templating all argocd helm charts and now pick them up via kustomize. Thats way less hassle in my opinion. I used [this](https://github.com/syncier/helmt) wrapper for helm template to make updating the charts easier

Appearently so.. I'm just astounded as to HOW argocd managed to break helm so profoundly :( - it has never failed me in the past.

**Comment by KlavsKlavsen** [comment]
@KarstenSiemer 
> using helm in argocd uses quite a lot of memory. Maybe it was just oomed either by your resource limit or by the node it is running on?

Yes thats how it started.. see #5903  - but now I've removed memory limit - to be certain that does not happen again - but ArgoCD won't recover (the helm applications just constantly fail sync).. And this worked perfectly for weeks now.. but when its suddenly "busy with a lot of work".. something goes bonkers with how argocd uses helm.. and logs don't offer a lot of help :(

**Comment by KlavsKlavsen** [comment]
unless its saying that Helm just died and didn't output anything? (as that seems to be what the log states)

**Comment by KlavsKlavsen** [comment]
"Manifest generation error (cached): `helm dependency build".. is hat what it logs now. WHO is caching this.. is this argocd "remembering" it failed last.. and not actually retrying.. or is it something else?

**Comment by KlavsKlavsen** [comment]
got it.. after it failing for "quite some time" - the UI offered me a "hard refresh" popdown on the refresh button (why isn't that shown all the time?) - and choosing that - fixed that application..
So perhaps the kill'ing of all pods actually DID resolve the issue - but ArgoCD was caching last result "indefinetely" - and never retrying?

It could seem this issue stems mostly from ArgoCD caching for too long (no cache timeout for failures appearently) ?

**Comment by KlavsKlavsen** [comment]
It seems the down arrow is just VERY HARD to hit precisely - and hence if oftens selects 'refresh' instead of showing the 'hard refresh' option :)

**Comment by KlavsKlavsen** [comment]
Every application is now going OK when I choose 'hard refresh' manually.
The fix for this would seem to be something like:
1) ensure "Helm result cache" is NOT long lived.. "positive cache" may live as long as git repo has not changed, but for sync - negative results should NOT live for more than a few minutes at max as they seem to hit some Helm issue which then needs manually hard refresh on every application.
2) Helm appearently messes up its cache on times.. so on negative Helm results - you should probably do this inside the pod:
rm -rf ~/.helm/cache/archive/*
rm -rf ~/.helm/repository/cache/*
helm repo update

which is what I "get" by killing pods, when changing replicas.

**Comment by KlavsKlavsen** [comment]
I would wager a guess - that argocd runs multiple Helm instances simultaneously inside same pod? That could explain how this Helm cache breakage happens - and locking around Helm "runner" - should resolve this. (and perhaps an upstream helm issue about multiple helm instances using same cache and corrupting it might help fix the issue :)

**Comment by KlavsKlavsen** [comment]
After having hard-refreshed a few applications - ArgoC appearently "unstuck" itself.. so the "helm cache" seems to be "shared amongst all applications" ? - and hence one hard-refresh "per repoServer pod" seems to resolve it ?

**Comment by KlavsKlavsen** [comment]
To anyone hitting this - I can reliably resolve "unknown sync status" - by scaling replica's for repo-server deploy down to 0 and up again - and then choose "Hard refresh" - by clicking arrow down on refresh.

**Comment by KlavsKlavsen** [comment]
It also seems as if "helm repo add" is not cached anywhere.. my application sync checks is starting to fail with the message: "helm repo add stable https://charts.helm.sh/stable` failed signal: killed"
It really should not need to add repos on EVERY helm run, but it seems it does?

**Comment by dorr-vonage** [comment]
Just had the same issue. Tried above mentioned workaround,  none worked. the repoServer log kept complaining that the dependency repo needs to be added.
Once I removed the Chart.lock file, all worked. No need to register the repo or define the controller or the repoServer options.
I'm using argocd v1.8.7

**Comment by KlavsKlavsen** [comment]
Thats not the same issue then.. What was your exact error message? does not sounds like the one above at all. The above is because it tries to add all helm repos - on every run - and that appearently fails (and stable is the first helm repo in my repositories list).

**Comment by dorr-vonage** [comment]
right, had too many open tabs and added the comment in the wrong one

**Comment by Talador12** [comment]
+1 on this issue

**Comment by pnowy** [comment]
I'm working with ArgoCD quite long time but notice this problem first time. The issue is that usually when I'm using external dependencies I don't have to put the chart in git repo. Example `Chart.yaml` (which is pointed in ArgoCD as main app):

```
apiVersion: v2
name: monitoring

dependencies:
  - name: kube-prometheus-stack
    version: 36.2.0
    repository: https://prometheus-community.github.io/helm-charts # https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
```

So this is `kube-prometheus-stack` chart which has other dependencies inside to prometheus-operator, grafana, etc. and working without any issues (right now on Argo 2.4.X but haven't seen any issues in previous ArgoCD versions).

I notice a problem today with our in-house chart (which is delivered by OCI repository so maybe that's the issue). In our chart we have optional dependency to grafana chart. Chart definition looks like (very similar to the previous one):

```
apiVersion: v2
name: myapp

dependencies:
  - name: myapp
    version: X.Y.Z
    repository: oci://repo/helm
```

The `myapp` when rendered on ArgoCD failed on the grafana dependencies rendering part.

For example when rendering: `{{ include "grafana.fullname" .Subcharts.grafana }}` ArgoCD complaining that tpl `grafana.fullname` doesn't exist at all (and my current ArgoCD version has 3.8.1 Helm version so subchart template should be supported as it was introduced in 3.7.X version).

To fix the problem I had to make locally `helm dep update` for myapp helm chart and commit tar.gz directly.

Does anyone had similar issue?

**Comment by dadav** [comment]
Also having this problem in the latest OpenShift version + gitops operator.

**Comment by migueleliasweb** [comment]
I've been hit by this issue as well. Interestingly, if I run my template locally with `helm template --dependency-update` it works as expected. Note that I'm purposedly not pushing the `charts/` folder with my git repo.

I've found this code in https://github.com/argoproj/argo-cd/blob/master/reposerver/repository/repository.go#L1083:

```golang
	out, err := h.Template(templateOpts)
	if err != nil {
		if !helm.IsMissingDependencyErr(err) {
			return nil, err
		}

		if concurrencyAllowed {
			err = runHelmBuild(appPath, h)
		} else {
			err = h.DependencyBuild()
		}

		if err != nil {
			return nil, err
		}

		out, err = h.Template(templateOpts)
		if err != nil {
			return nil, err
		}
	}
```

It specifically checks for missing deps and it `should` run `helm dependency update` but for some reason this doesn't seem to be doing much.

Any thoughts?

**Comment by lebenitza** [comment]
@dcandas I think your use case is caused by this issue: https://github.com/helm/helm/issues/2247 
I just ran into this as well but I don't think it is an ArgoCD problem

**Comment by davhdavh** [comment]
I am surprised noone added this feature, as far as I can tell it should be pretty simple: 
1. ensure this can parse file:// properly: https://github.com/argoproj/argo-cd/blob/a1f90b5cb6d3b5a2d642a0cfe0b2da2f6bd89342/reposerver/repository/repository.go#L1019
2. update this to recursively call itself on all file:// directory references https://github.com/argoproj/argo-cd/blob/a1f90b5cb6d3b5a2d642a0cfe0b2da2f6bd89342/reposerver/repository/repository.go#L1089
3. add some sort of loop detection or a max-depth

**Comment by sarendipitee** [comment]
So what's the recommended way of using helm charts then? Or is using helm charts at all discouraged?

**Comment by rumstead** [comment]
I deploy Helm dependencies with Argo all the time. This issue is from Argo CD 1.4.2. I am pretty sure this issue can be closed.

**Comment by rumstead** [comment]
The error message indicated that the lock file was out of sync with `requirements.yaml`, which prevented manifest generation. The solution is to run helm dependency build to ensure the lock file is up to date before pushing the chart to the repository to allow Argo CD can deploy them as expected. Closing this issue now to document the resolution for others who may encounter it.

If anyone faces a similar problem, please open a new issue.

**Comment by dntosas** [comment]
> The error message indicated that the lock file was out of sync with `requirements.yaml`, which prevented manifest generation. The solution is to run helm dependency build to ensure the lock file is up to date before pushing the chart to the repository to allow Argo CD can deploy them as expected. Closing this issue now to document the resolution for others who may encounter it.
> 
> If anyone faces a similar problem, please open a new issue.

this is not solved @rumstead . there is no `requirements.yaml` concept in Helm and is happening on Argo v3.1.0

**Comment by dntosas** [comment]
@rumstead this function does not parse `file://..` type https://github.com/argoproj/argo-cd/blob/a1f90b5cb6d3b5a2d642a0cfe0b2da2f6bd89342/reposerver/repository/repository.go#L1019

**Comment by kresimir-ostovic** [comment]
still happening version 2.14.11
Is @rumstead some llm?

**Comment by rumstead** [comment]
No I am not an LLM :) but please create a new issue. It is hard for maintainers to disambiguate the original issue that was reported and comments years later.
